### ISSUE TITLE
Travis CI is awfully slow

### ISSUE THREAD

**Issue by untitaker** [untitaker/vdirsyncer#362]
With the current amount of servers tested, CI takes **two and a half hours** to complete.

I'm not sure how to get around this, but the solution I'm thinking about involves *two* CI services, one that only tests `make DAV_SERVER=radicale test` and `make style`, and one that test the whole range. The second one would be called only for master or so.

I'm aware of the following CI services that can be integrated with GitHub:

* https://drone.io/
* https://circleci.com/
* Travis itself

For the full build, we absolutely need root access, and the ability to add arbitrary third-party PPAs. As far as I understand, this is a feature that Travis, Drone and Circle CI all have.

A setup using two services is pretty unfortunate, however. I'd prefer a solution that allows one to define a different build matrix for PRs than for pushes to master or tags.

--

This supersedes #350 and probably depends on #361

**Comment by hobarrera** [comment]
Testing against externally-running servers would reduce test times, since we wouldn't be doing all the installation part.

We could also reduce the matrix a bit. I don't think we need to test every python version vs bakai, we can test just 2.7 and 3.5. The same goes for other servers.
As long as we run all the tests for each python version against ONE server, we might be covered.

**Comment by untitaker** [comment]
>Testing against externally-running servers would reduce test times, since we wouldn't be doing all the installation part.

One thing I want to keep with the current setup is reproducability: It should be possible to run tests locally for most of them. Davical is the exception.

Hopefully there's a CI out there that provides some sort of caching.

>I don't think we need to test every python version vs bakai, we can test just 2.7 and 3.5.

This is already the case. Only Radicale is tested on all versions.

**Comment by hobarrera** [comment]
Do we really have reproducibility right now? If one of the testserver repos is updated, and old passing test might break (if the server broke something). We're not actually pinning server versions to each commit either.

**Comment by untitaker** [comment]
Probably the wrong word, but at least I can assure myself that the bugs are real and not caused by Travis. Offline testing is really the thing I care about.

**Comment by untitaker** [comment]
Great, I broke the master build horribly because I was too impatient for Travis to finish.

**Comment by untitaker** [comment]
I've just added http://homu.io/ to this repo. It implements a PR merging queue: Basically you do `@homu r+` and the bot will test the PR and automatically merge it after CI passes.

Unsure if this model a good idea, but it's successfully used by several Rust projects.

**Comment by hobarrera** [comment]
I like the fact that homu merges after the tests completes, especially since tests take a while. It' works really well for our rather long tests.

(OT: GitLab has this out-of-the-box ;) )

**Comment by untitaker** [comment]
So the current workflow for testing is:

* If you open a PR, the tests will not be automatically run. No status at all.
* If a [reviewer](http://homu.io/r/untitaker/vdirsyncer) comments `@homu r+`, homu does its thing:

  * Force push the `auto` branch to have the same HEAD as `master`
  * Push merge commit to the `auto` branch
  * Push to master if status succeeds

As of #376, `auto` is the only branch Travis tests.

I'd now like to install another CI service that will run minimal tests (stylecheck and basic testsuite) when the PR is opened. Ideally this suite should finish under 5 minutes such that the contributor gets early feedback (something that is missing when running the full testsuite).

**Comment by untitaker** [comment]
@hobarrera And yes, GitLab does the merging queue thing, but having different build matrices for PRs and master is way out of scope for either GitHub or GitLab. :)

**Comment by hobarrera** [comment]
Yep, it looks like travis won't allow this. GitLab actually allows defining branches (or tags) on a per-task basis.

**Comment by untitaker** [comment]
>GitLab actually allows defining branches (or tags) on a per-task basis.

I think we need the inverse: Defining tasks/matrix on a per-branch basis.

**Comment by untitaker** [comment]
Discard previous comments. I found http://doc.gitlab.com/ce/ci/yaml/README.html, and it's exactly what we need.

Meanwhile Travis has rejected this feature request [two years ago](https://github.com/travis-ci/travis-ci/issues/2778), but it doesn't seem that they thought of our usecase. Maybe they will change their mind.

**Comment by untitaker** [comment]
Drone.io and CircleCI are not usable.

CircleCI is way too presumptious regarding configuration, and tries to guess *everything* (the programming language, how to setup and run tests), resulting in unbelievable mess. The only thing it correctly achieved was installing vdirsyncer in a virtualenv. Furthermore it doesn't offer a proper way to delete the project or my account, which is shady to say the least.

Drone.io is easy to use, but also very primitive. It only offers one job, and the latest Python 3 there is Python 3.3. While it integrates with GitHub, Bitbucket and Google Code, it offers virtually no integration specific to those platforms, i.e. support for GitHub's commit status is missing.

**Comment by untitaker** [comment]
I think the next-best thing what we can do is to simply make most of the jobs a no-op for PRs, as documented in https://docs.travis-ci.com/user/customizing-the-build/#Implementing-Complex-Build-Steps.

Shame on Travis, but I guess it's their money we're pulverizing.

**Comment by hobarrera** [comment]
> Shame on Travis, but I guess it's their money we're pulverizing.

GitLab *does* have a pretty long queue at times too, so none of them give us lots of free hardware permanently anyway. Upside is, you can add your own worker machines (which I have, since I need them for work).

Maybe we need to deal with the fact that PRs are slow, and make the best of homu. :(

**Comment by untitaker** [comment]
> GitLab *does* have a pretty long queue at times too, so none of them give us
> lots of free hardware permanently anyway. Upside is, you can add your own
> worker machines (which I have, since I need them for work).

I find it ironic that Travis encourages users to avoid excessive usage of their
hardware through caching, container-based builds, but then gives me no other
choice other than to start 30+ builds for nothing.

I didn't mean to sound entitled to free ressources.

> Maybe we need to deal with the fact that PRs are slow, and make the best of homu. :(

Probably. I hope this will get better once we fix #232
